### PR TITLE
Docstring check

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds a new util script that checks docstrings documentation of the arguments match the signature. This first version only checks base objects (functions and classes) not class methods. It checks:
- that the defaults documented are the same as the signature
- applies our traditional formatting (numbers not in code blocks, all other objects in code blocks, default to None skipped)
- that the order of the parameters is the same in the docstring and in the signature
- that all arguments (except `*args` or `**kwargs` or private arguments) are documented

It also performs auto-fixing for all of those problems (with templates for arguments that are not documented). As you will see from the diff (and it only covers the objects where all arguments were documented!) this was direly needed and should help a lot in the future.

The auto-fix part is added in `make fix-copies` while the check is in `make repo-consistency`. Since the auto fixing adds templates sometimes (for non-documented arguments), the check will error if those templates haven't been replaced, so the CI tells us when something is wrong.

The script has been applied to all objects it could fully auto-fix, and there is a (long) list of objects where manual input is needed (because some arguments are not documented yet). This could be fixed progressively with a sprint involving the community (hacktoberfest is coming!), but at least we can ensure newly added objects are properly documented.

A couple of additional notes:
- this is only applied to objects that are in the public init and not ignored in the check that ensures all objects are documented.
- as said before this doesn't touch the method yet, only the class inits. Adding the methods will be done in a followup PR later on
- one can ignore a docstring from this check by adding a comment `# no-format` before the docstring (see for instance `TrainerCallback`)
- it is also possible to ignore the re-ordering of the arguments documentation (if there is a special order that makes more sense for the documentation) by adding a comment `# ignore-order` before the docstring (no example of this yet)
- since we have a lot of arguments with a default `None` in the signature but that are set to a value in the call (often for mutable arguments since default mutables are dangerous in Python) the check won't replace a documented default when the signature shows None.
- we also have a lot of default arguments that are defined by mathematical formulas, which make more sense than the actual value, so the check uses a small Python interpreter for math expressions to make sure to leave those alone (if they are accurate).
- some values are reformatted by Python: it wants `1e-05` instead of `1e-5` or `Resampling.Xxx` instead of `PILResampling.Xxx`. Those are minor changes we don't really care about (I think).

cc @amyeroberts since you were interested by this
cc the doc master @stevhliu and @MKhalusova for your information

### CODE REVIEW

File: src/transformers/models/clap/feature_extraction_clap.py
Comment (ArthurZucker) [inline src/transformers/models/clap/feature_extraction_clap.py:1]: oups 😅

File: src/transformers/models/lxmert/configuration_lxmert.py [hunk 1]
```diff
@@ -69,25 +73,21 @@ class LxmertConfig(PretrainedConfig):
...
+        r_layers (`int`, *optional*, defaults to 5):
+            Number of hidden layers in the Transformer visual encoder.
         visual_feat_dim (`int`, *optional*, defaults to 2048):
             This represents the last dimension of the pooled-object features used as input for the model, representing
             the size of each object feature itself.
         visual_pos_dim (`int`, *optional*, defaults to 4):
             This represents the number of spacial features that are mixed into the visual features. The default is set
             to 4 because most commonly this will represent the location of a bounding box. i.e., (x, y, width, height)
-        visual_loss_normalizer (`float`, *optional*, defaults to 1/15):
+        visual_loss_normalizer (`float`, *optional*, defaults to 6.67):
```
Comment (ArthurZucker) [inline src/transformers/models/lxmert/configuration_lxmert.py:88]: The next file keeps it as `1/15`

File: src/transformers/models/mask2former/image_processing_mask2former.py [hunk 2]
```diff
@@ -356,20 +356,17 @@ class Mask2FormerImageProcessor(BaseImageProcessor):
...
             `PIL.Image.Resampling.BOX`, `PIL.Image.Resampling.BILINEAR`, `PIL.Image.Resampling.HAMMING`,
             `PIL.Image.Resampling.BICUBIC` or `PIL.Image.Resampling.LANCZOS`. Only has an effect if `do_resize` is set
             to `True`.
-        size_divisor (`int`, *optional*, defaults to 32):
-            Some backbones need images divisible by a certain number. If not passed, it defaults to the value used in
-            Swin Transformer.
         do_rescale (`bool`, *optional*, defaults to `True`):
             Whether to rescale the input to a certain `scale`.
-        rescale_factor (`float`, *optional*, defaults to 1/ 255):
+        rescale_factor (`float`, *optional*, defaults to `1/ 255`):
```
Comment (ArthurZucker) [inline src/transformers/models/mask2former/image_processing_mask2former.py:369]: Seem like it should be converted to the float format? (above file) 
let's settle maybe

File: src/transformers/models/nat/configuration_nat.py [hunk 3]
```diff
@@ -65,7 +65,7 @@ class NatConfig(BackboneConfigMixin, PretrainedConfig):
             `"selu"` and `"gelu_new"` are supported.
         initializer_range (`float`, *optional*, defaults to 0.02):
             The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
-        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
+        layer_norm_eps (`float`, *optional*, defaults to 0.0):
```
Comment (ArthurZucker) [inline src/transformers/models/nat/configuration_nat.py:68]: this is wrong should be `1e-05`

File: src/transformers/models/pvt/configuration_pvt.py [hunk 4]
```diff
@@ -49,7 +49,7 @@ class PvtConfig(PretrainedConfig):
             The input image size
         num_channels (`int`, *optional*, defaults to 3):
             The number of input channels.
-        num_encoder_blocks (`[int]`, *optional*., defaults to 4):
+        num_encoder_blocks (`[int]`, *optional*, defaults to 4):
```
Comment (ArthurZucker) [inline src/transformers/models/pvt/configuration_pvt.py:52]:
```suggestion
        num_encoder_blocks (`int`, *optional*, defaults to 4):
```
something seems a bit wrong with this one

File: src/transformers/models/pvt/configuration_pvt.py
Comment (ArthurZucker) [inline src/transformers/models/pvt/configuration_pvt.py:1]: Good example here the `hidden_act`  is not properly detected (it's Mapping: `hidden_act: Mapping[str, Callable] = "gelu",`

Comment (ArthurZucker) [reply src/transformers/models/pvt/configuration_pvt.py:1]: But I guess it's pretty specific / special

File: src/transformers/models/swin/configuration_swin.py [hunk 5]
```diff
@@ -74,13 +74,13 @@ class SwinConfig(BackboneConfigMixin, PretrainedConfig):
         hidden_act (`str` or `function`, *optional*, defaults to `"gelu"`):
             The non-linear activation function (function or string) in the encoder. If string, `"gelu"`, `"relu"`,
             `"selu"` and `"gelu_new"` are supported.
-        use_absolute_embeddings (`bool`, *optional*, defaults to False):
+        use_absolute_embeddings (`bool`, *optional*, defaults to `False`):
             Whether or not to add absolute position embeddings to the patch embeddings.
         initializer_range (`float`, *optional*, defaults to 0.02):
             The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
-        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
+        layer_norm_eps (`float`, *optional*, defaults to 0.0):
```
Comment (ArthurZucker) [inline src/transformers/models/swin/configuration_swin.py:81]:
```suggestion
        layer_norm_eps (`float`, *optional*, defaults to 1e-05):
```

Comment (ArthurZucker) [reply src/transformers/models/swin/configuration_swin.py:81]: seems like this is recurrent with the very very small default

File: src/transformers/models/swin2sr/configuration_swin2sr.py [hunk 6]
```diff
@@ -69,12 +69,12 @@ class Swin2SRConfig(PretrainedConfig):
             Whether or not to add absolute position embeddings to the patch embeddings.
         initializer_range (`float`, *optional*, defaults to 0.02):
             The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
-        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
+        layer_norm_eps (`float`, *optional*, defaults to 0.0):
```
Comment (ArthurZucker) [inline src/transformers/models/swin2sr/configuration_swin2sr.py:72]:
```suggestion
        layer_norm_eps (`float`, *optional*, defaults to 1e-05):
```

File: src/transformers/models/swinv2/configuration_swinv2.py [hunk 7]
```diff
@@ -70,9 +70,9 @@ class Swinv2Config(PretrainedConfig):
             Whether or not to add absolute position embeddings to the patch embeddings.
         initializer_range (`float`, *optional*, defaults to 0.02):
             The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
-        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
+        layer_norm_eps (`float`, *optional*, defaults to 0.0):
```
Comment (ArthurZucker) [inline src/transformers/models/swinv2/configuration_swinv2.py:73]:
```suggestion
        layer_norm_eps (`float`, *optional*, defaults to 1e-05):
```

File: src/transformers/trainer_utils.py [hunk 8]
```diff
@@ -111,7 +111,7 @@ class EvalPrediction:
     Parameters:
         predictions (`np.ndarray`): Predictions of the model.
         label_ids (`np.ndarray`): Targets to be matched.
-        inputs (`np.ndarray`, *optional*)
+        inputs (`np.ndarray`, *optional*):
```
Comment (ArthurZucker) [inline src/transformers/trainer_utils.py:114]: Should fail on this no? saying no doc and prediction and labels_ids shoud return to new line

File: src/transformers/utils/quantization_config.py [hunk 9]
```diff
@@ -167,7 +167,7 @@ class BitsAndBytesConfig(QuantizationConfigMixin):
         bnb_4bit_compute_dtype (`torch.dtype` or str, *optional*, defaults to `torch.float32`):
             This sets the computational type which might be different than the input time. For example, inputs might be
             fp32, but computation can be set to bf16 for speedups.
-        bnb_4bit_quant_type (`str`, {fp4, nf4}, defaults to `fp4`):
+        bnb_4bit_quant_type (`str`, {fp4, nf4}, defaults to `fp4`, *optional*, defaults to `"fp4"`):
```
Comment (ArthurZucker) [inline src/transformers/utils/quantization_config.py:170]: this is a very edgy case
```suggestion
        bnb_4bit_quant_type (`str`,  *optional*, defaults to `"fp4"`):
```

File: tests/repo_utils/test_check_docstrings.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+import unittest
+
+
+git_repo_path = os.path.abspath(os.path.dirname(os.path.dirname(os.path.dirname(__file__))))
+sys.path.append(os.path.join(git_repo_path, "utils"))
+
+from check_docstrings import get_default_description, replace_default_in_arg_description  # noqa: E402
+
+
+class CheckDostringsTested(unittest.TestCase):
```
Comment (ArthurZucker) [inline tests/repo_utils/test_check_docstrings.py:27]: Few more edge cases can be added from the comments I left!

File: utils/check_docstrings.py [hunk 11]
```diff
@@ -0,0 +1,1266 @@
+# coding=utf-8
+# Copyright 2020 The HuggingFace Inc. team.
```
Comment (ArthurZucker) [inline utils/check_docstrings.py:2]:
```suggestion
# Copyright 2023 The HuggingFace Inc. team.
```

File: utils/check_docstrings.py [hunk 12]
```diff
@@ -0,0 +1,1266 @@
...
+# This is a temporary list of objects to ignore while we progressively fix them. Do not add anything here, fix the
+# docstrings instead. If formatting should be ignored for the docstring, you can put a comment # no-format on the
+# line before the docstring.
+OBJECTS_TO_IGNORE = [
+    # Deprecated
+    "InputExample",
+    "InputFeatures",
+    # Signature is *args/**kwargs
+    # "PretrainedConfig", #ignored but could be fixed
+    # "GenerationConfig", #ignored but could be fixed
```
Comment (ArthurZucker) [inline utils/check_docstrings.py:72]:
```suggestion
```

File: utils/check_docstrings.py [hunk 13]
```diff
@@ -0,0 +1,1266 @@
...
+    return len(search.groups()[0])
+
+
+def stringify_default(default: Any) -> str:
+    """
+    Returns the string representation of a default value, as used in docstring: numbers are left as is, all other
+    objects are in backtiks.
+
+    Args:
+        default (`Any`): The default value to process
```
Comment (ArthurZucker) [inline utils/check_docstrings.py:862]:
```suggestion
        default (`Any`): 
            The default value to process
```

File: utils/check_docstrings.py [hunk 14]
```diff
@@ -0,0 +1,1266 @@
...
+        return f"`{default}`"
+
+
+def eval_math_expression(expression: str) -> Optional[Union[float, int]]:
+    # Mainly taken from the excellent https://stackoverflow.com/a/9558001
+    """
+    Evaluate (safely) a mathematial expression and returns its value.
+
+    Args:
+        expression (`str`): The expression to evaluate.
```
Comment (ArthurZucker) [inline utils/check_docstrings.py:891]:
```suggestion
        expression (`str`): 
            The expression to evaluate.
```

**Review by ArthurZucker — COMMENTED** [review]
Is it Christma? 🎁 
Thanks a mile for this utility! 
Mostly reviewed each individual files for now, found 1-2 edge cases but otherwise LGTM! Will review again the checker!

File: src/transformers/models/mask2former/image_processing_mask2former.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/models/mask2former/image_processing_mask2former.py:369]: When it's a formula, it is put in code style.

File: src/transformers/models/nat/configuration_nat.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/models/nat/configuration_nat.py:68]: Crap, it's the rounding of floats that does that. Will have a look when I can.

File: src/transformers/utils/quantization_config.py (hunk 9 above)
Comment (sgugger) [reply src/transformers/utils/quantization_config.py:170]: Yes in those cases the script does that (I can't add the optional at the right place) but it should be easy to spot in reviews.

File: utils/check_docstrings.py (hunk 12 above)
Comment (sgugger) [reply utils/check_docstrings.py:72]: No I want to leave them here in the list of objects to try to fix.

File: utils/check_docstrings.py (hunk 13 above)
Comment (sgugger) [reply utils/check_docstrings.py:862]: Both syntaxes are acceptable ;-p

File: utils/check_docstrings.py (hunk 14 above)
Comment (sgugger) [reply utils/check_docstrings.py:891]: Same here.

File: src/transformers/trainer_utils.py (hunk 8 above)
Comment (sgugger) [reply src/transformers/trainer_utils.py:114]: It's hard to detect there is no actual documentation, I think that's something the reviewer can spot easily.
As for return to new lines, it's only necessary for long docstrings, the syntax for predictions and label_ids is perfectly fine and accepted by the doc builder.

File: src/transformers/models/lxmert/configuration_lxmert.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/models/lxmert/configuration_lxmert.py:88]: Here the default in the signature is 6.67, so not sure where this comment is coming from?

File: src/transformers/models/pvt/configuration_pvt.py
Comment (sgugger) [reply src/transformers/models/pvt/configuration_pvt.py:1]: This does not check the type annotations, just the defaults.

**Review by ydshieh — COMMENTED** [review]
Thank you a lot, this is going to ease much our life ❤️ !

I have limited time (before I depart next week), but one question on my side:

**what happens if the docstring contains argument name(s) that are not signature?**

It seems this is not treated explicitly. And see my comment below in `check_docstring.py`.

I can take a look more thoughly when I'm back (if it is not merge at that time then).

File: utils/check_docstrings.py [hunk 15]
```diff
@@ -0,0 +1,1267 @@
...
+            inspect._ParameterKind.VAR_POSITIONAL,
+        ]:
+            arguments[name] = ""
+        else:
+            arg_desc = get_default_description(arg)
+            arguments[name] = " " * (indent + 4) + f"{name} ({arg_desc}): <fill_docstring>"
+
+    # Arguments are sorted by the order in the signature unless a special comment is put.
+    if ignore_order:
+        new_param_docs = [arguments[name] for name in old_arguments]
```
Comment (ydshieh) [inline utils/check_docstrings.py:1133]: this includes all order arguments + the missing one.

File: utils/check_docstrings.py [hunk 16]
```diff
@@ -0,0 +1,1267 @@
...
+        missing = set(signature.keys()) - set(old_arguments)
+        new_param_docs.extend([arguments[name] for name in missing if len(arguments[name]) > 0])
+    else:
+        new_param_docs = [arguments[name] for name in signature.keys() if len(arguments[name]) > 0]
```
Comment (ydshieh) [inline utils/check_docstrings.py:1137]: this include only those in `signature`.

File: utils/check_docstrings.py (hunk 15 above)
Comment (sgugger) [reply utils/check_docstrings.py:1133]: Right, here the arguments not in the signature should be removed. Will add that!

**Comment by sgugger** [comment]
> what happens if the docstring contains argument name(s) that are not signature?

Then the argument in the docstrings should be removed. The goal is to ensure consistency, and arguments not in the signature are often hidden for a reason (deprecated arguments for instance) or might even not be accepted (but end up in the docstring after a bad copy paste for instance). Will fix the behavior you flagged in a comment.

**Comment by ArthurZucker** [comment]
Feel free to ping me again for another review!

**Comment by sgugger** [comment]
Waiting for @ydshieh review when he is back from vacation.

**Comment by ydshieh** [comment]
Here is a second review where I found something to check:

- Previous `PILImageResampling.BICUBIC` now becomes `Resampling.BICUBIC` (beit image processor etc), similar to `Resampling.BILINEAR` (convnext image processor etc). Do you know why?
- `**optional**` is changed to `**optional*`, but I think we want `*optional*` (in logit processor file)
- ``["<s>NOTUSED", "</s>NOTUSED"]`` become `['<s>NOTUSED', '</s>NOTUSED']` (list of str):  I am fine with it, but want to double check with you.
- There is `eps: int = 1e-6,` but I think type annotation is not the goal of this PR

**Comment by sgugger** [comment]
Yes, the tool does not check for type annotations. I can try to catch the `**optional**` floating around, but the tool wasn't intended to fix this either, the modifications you see are just side effects.

As for the two other points, this falls under the last point of the description: some values are reformatted, but the tool will be consistent so I don't think it's a bad thing, it's just a different thing.

Let me know if you think of other things to add.

File: src/transformers/generation/logits_process.py [hunk 17]
```diff
@@ -1600,18 +1600,15 @@ class UnbatchedClassifierFreeGuidanceLogitsProcessor(LogitsProcessor):
...
+        unconditional_attention_mask (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
             Attention mask for unconditional_ids.
-        model (`PreTrainedModel`):
-            The model computing the unconditional scores. Supposedly the same as the one computing the conditional
-            scores. Both models must use the same tokenizer.
-        smooth_factor (`float`, **optional**):
-            The interpolation weight for CFG Rescale. 1 means no rescaling, 0 reduces to the conditional scores without
-            CFG. Turn it lower if the output degenerates.
-        use_cache (`bool`, **optional**):
+        use_cache (`bool`, **optional*, defaults to `True`):
```
Comment (ydshieh) [inline src/transformers/generation/logits_process.py:1611]: There is this remaining one :-)

**Comment by ydshieh** [comment]
It turns out `PILImageResampling` is not from PIL but our internal code. So good to keep  `PILImageResampling`.

**Review by ydshieh — APPROVED** [review]
Thank you again for this awesome work, love it ❤️ .

Just a final `**optional*` thing.

File: src/transformers/models/deprecated/mctct/configuration_mctct.py [hunk 18]
```diff
@@ -80,17 +80,17 @@ class MCTCTConfig(PretrainedConfig):
             The probability of randomly dropping the `Conv1dSubsampler` layer during training.
         num_conv_layers (`int`, *optional*, defaults to 1):
             Number of convolution layers before applying transformer encoder layers.
-        conv_kernel (`List[int]`, *optional*, defaults to `[7]`):
+        conv_kernel (`List[int]`, *optional*, defaults to `(7,)`):
```
Comment (LysandreJik) [inline src/transformers/models/deprecated/mctct/configuration_mctct.py:83]: Haha it accepts `List[int]` but refuses to keep the list as the default

Comment (LysandreJik) [reply src/transformers/models/deprecated/mctct/configuration_mctct.py:83]: IIUC this doesn't change the expected types, only the default, so maybe worth updating the type hint here

File: src/transformers/models/deprecated/mctct/configuration_mctct.py [hunk 19]
```diff
@@ -80,17 +80,17 @@ class MCTCTConfig(PretrainedConfig):
...
             The kernel size of the 1D convolution applied before transformer layers. `len(conv_kernel)` must be equal
             to `num_conv_layers`.
-        conv_stride (`List[int]`, *optional*, defaults to `[3]`):
+        conv_stride (`List[int]`, *optional*, defaults to `(3,)`):
```
Comment (LysandreJik) [inline src/transformers/models/deprecated/mctct/configuration_mctct.py:86]:
```suggestion
        conv_stride (`Sequence[int]`, *optional*, defaults to `(3,)`):
```

File: src/transformers/models/deprecated/mctct/configuration_mctct.py (hunk 18 above)
Comment (LysandreJik) [reply src/transformers/models/deprecated/mctct/configuration_mctct.py:83]: Will GitHub refuses me to add a hint, but wanted to offer a change from `List` to `Sequence`

File: utils/check_docstrings.py [hunk 20]
```diff
@@ -0,0 +1,1272 @@
...
+# Re pattern that parses the start of an arg block: catches <name> (<description>) in those lines.
+_re_parse_arg = re.compile(r"^(\s*)(\S+)\s+\((.+)\)(?:\:|$)")
+# Re pattern that parses the end of a description of an arg (catches the default in *optional*, defaults to xxx).
+_re_parse_description = re.compile(r"\*optional\*, defaults to (.*)$")
+
+
+# This is a temporary list of objects to ignore while we progressively fix them. Do not add anything here, fix the
+# docstrings instead. If formatting should be ignored for the docstring, you can put a comment # no-format on the
+# line before the docstring.
+OBJECTS_TO_IGNORE = [
```
Comment (LysandreJik) [inline utils/check_docstrings.py:66]: Hacktoberfest :star_struck:

**Review by LysandreJik — APPROVED** [review]
Thanks for your PR!

When checking it out on my PC and running it I had a few issues. It seems to be quite finicky with dependencies: running it first identified 23 files that needed a change, mostly processors with the following changes:

```diff
diff --git a/src/transformers/models/beit/image_processing_beit.py b/src/transformers/models/beit/image_processing_beit.py
index 6f8ce403e..769624e67 100644
--- a/src/transformers/models/beit/image_processing_beit.py
+++ b/src/transformers/models/beit/image_processing_beit.py
@@ -57,7 +57,7 @@ class BeitImageProcessor(BaseImageProcessor):
         size (`Dict[str, int]` *optional*, defaults to `{"height": 256, "width": 256}`):
             Size of the output image after resizing. Can be overridden by the `size` parameter in the `preprocess`
             method.
-        resample (`PILImageResampling`, *optional*, defaults to `Resampling.BICUBIC`):
+        resample (`PILImageResampling`, *optional*, defaults to `3`):
             Resampling filter to use if resizing the image. Can be overridden by the `resample` parameter in the
             `preprocess` method.
         do_center_crop (`bool`, *optional*, defaults to `True`):
```

Reinstalling my environment resulted in these changes still passing the test (because `Resampling.BICUBIC == 3`). I don't think it's something that's simply solvable with your script (or maybe by ignoring values which cannot be read because they can't be imported/ignoring replacements that have the same value?), but I do think we should keep it in mind and expect future contributor PRs to have seemingly unrelated changes linked to them running the style changes.

Maybe we could eventually take a look at making it faster (runs in ~19 seconds on my machine)

**Comment by sgugger** [comment]
I think the discrepancy in values is due to the different versions of PIL, but I didn't manage to reproduce having 3 instead of the enum object printed.

For speed, it's a bit challenging to have the util only verify the files modified (like `make fixup`) instead of everything, since it's not based on file but on objects. Part of the slowness is also just importing everything from `transformers`.

**Comment by LysandreJik** [comment]
Sounds good! Let's merge this PR.

Thanks for your efforts @sgugger!

**Comment by LysandreJik** [comment]
Merged! Thanks 🫡

**Pull request merged**